homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,449 @@
         try:
-            keep_alive(None)
-            # Loop through every device registered with the iCloud account
-            for device in api.devices:
-                status = device.status()
-                dev_id = slugify(status['name'].replace(' ', '', 99))
-
-                # An entity will not be created by see() when track=false in
-                # 'known_devices.yaml', but we need to see() it at least once
-                entity = hass.states.get(ENTITY_ID_FORMAT.format(dev_id))
-                if entity is None and dev_id in seen_devices:
-                    continue
-                seen_devices[dev_id] = True
-
-                location = device.location()
-                # If the device has a location add it. If not do nothing
-                if location:
-                    see(
-                        dev_id=dev_id,
-                        host_name=status['name'],
-                        gps=(location['latitude'], location['longitude']),
-                        battery=status['batteryLevel']*100,
-                        gps_accuracy=location['horizontalAccuracy']
-                    )
+            for device in self.api.devices:
+                if str(device) == str(self.devices[devicename]):
+                    status = device.status(DEVICESTATUSSET)
+                    dev_id = status['name'].replace(' ', '', 99)
+                    dev_id = slugify(dev_id)
+                    attrs[ATTR_DEVICESTATUS] = DEVICESTATUSCODES.get(
+                        status['deviceStatus'], 'error')
+                    attrs[ATTR_LOWPOWERMODE] = status['lowPowerMode']
+                    attrs[ATTR_BATTERYSTATUS] = status['batteryStatus']
+                    attrs[ATTR_ACCOUNTNAME] = self.accountname
+                    status = device.status(DEVICESTATUSSET)
+                    battery = status.get('batteryLevel', 0) * 100
+                    location = status['location']
+                    if location:
+                        self.determine_interval(
+                            devicename, location['latitude'],
+                            location['longitude'], battery,
+                            attrs.get(ATTR_GMTT_DURATION))
+                        interval = self._intervals.get(devicename, 1)
+                        attrs[ATTR_INTERVAL] = interval
+                        accuracy = location['horizontalAccuracy']
+                        kwargs['dev_id'] = dev_id
+                        kwargs['host_name'] = status['name']
+                        kwargs['gps'] = (location['latitude'],
+                                         location['longitude'])
+                        kwargs['battery'] = battery
+                        kwargs['gps_accuracy'] = accuracy
+                        kwargs[ATTR_ATTRIBUTES] = attrs
+                        self.see(**kwargs)
+                        self.seen_devices[devicename] = True
         except PyiCloudNoDevicesException:
```

Comment:
When can this exception occur? Can you limit the code inside the `try…catch` to only the piece of code that can raise.